sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala

```diff
@@ -674,4 +674,40 @@ class JsonFunctionsSuite extends QueryTest with SharedSparkSession {
       spark.range(1).select(schema_of_json(input)),
       Seq(Row("struct<id:bigint,price:double>")))
   }
+
+  test("SPARK-31065: schema_of_json - null and empty strings as strings") {
+    Seq("""{"id": null}""", """{"id": ""}""").foreach { input =>
+      checkAnswer(
+        spark.range(1).select(schema_of_json(input)),
+        Seq(Row("struct<id:string>")))
```

Comment:
In these days, the behavior change becomes a tricky issue although it's a bug fix. I'm wondering if this PR is safe (`not a silent behavior change`) in this case because Apache Spark 2.4.x ~ 3.0.0-preview2 returns like the following as mentioned in the PR description. To make it sure,  ping @gatorsmile , @cloud-fan , @marmbrus , @srowen .
```scala
scala> spark.range(1).select(schema_of_json("""{"id": null}""")).show
+----------------------------+
|schema_of_json({"id": null})|
+----------------------------+
|             struct<id:null>|
+----------------------------+
```

Reply:
From what I read here, the OP issue looks like a clean bug fix. Yes a behavior change but bug fixes are. The change you highlight here is subtler, yes. I think it's reasonable to infer string type rather than null type, but would only do it at 3.0, not Spark 2.x. It would be a release notes item.

Reply:
Yep. I agree.

Reply:
@dongjoon-hyun, the problem here is that `struct<id:null>` can't be used as a schema. The main purpose of `schema_of_json` is an alternative for schema inference for `from_json`.

Currently, the codes you mentioned don't work with it:

```scala
val schemaExpr = schema_of_json(lit("""{"id": ""}"""))
spark.range(1).select(from_json(lit("""{"id": ""}"""), schemaExpr)).show()
```

**Before:**

```
org.apache.spark.sql.catalyst.parser.ParseException:
...
== SQL ==
struct<id:null>
------^^^
```

**After:**

```
+---------------------+
|from_json({"id": ""})|
+---------------------+
|                   []|
+---------------------+
```

`struct<id:null>` can't be used anywhere as we know `null` isn't supported as DDL formatted string.

It's unlikely users depend on this behaviour so I personally don't think it's worthwhile to add a configuration and I would even doubt about the release note if this fix only lands to Spark 3.0.

Reply:
Thanks. Got it. It makes sense to me, too.